### ISSUE TITLE
Account summary alongside the transactions data

### ISSUE THREAD

**Issue by chilikla** [eshaham/israeli-bank-scrapers#76]
My use case, for now, of using this library is getting all of my finance information in one place. For this matter, transactions data is less important for me, but accounts summaries are more useful. e.g. getting the current balance of my bank account(s), or knowing the my expected credit card(s) charge for this month.
Is this something this project willing to include?
I already wrote a code for myself for Bank Hapoalim and Leumi Card with new properties of 'balance' (in the case of the bank) and 'charge' (in the case of the credit card), in the returned object (alongside the success and txns properties).

In a related manner, what about investing data?
I have an Intergemel account, and I'm finding it useful to get data of "what my money did today" for it. I can contribute code for handling Intergemel accounts, with basic information about the type of account, its number, its nick name and its current balance of each of the customer accounts.

**Comment by eshaham** [comment]
Hey @chilikla 😄
I think that adding account summaries would be a great add-on to existing data we scrape.
can you suggest the signature changes for the existing scrape result?
I wonder if we could still maintain the same signature for both banks and credit cards when it comes to summaries. Something tells me it's going to be difficult.
We would need to add these to all scrapers of course, so it will be a joint effort.

As for Intergemel, I'm thinking it may not fit here. I've actually started another project and started scraping insurance/pension companies, but it is in a very initial stage:
https://github.com/eshaham/israeli-insurance-scrapers
see clalbit branch.

**Comment by chilikla** [comment]
Hi @eshaham :)
I don't really understand what do you mean by maintaining same signature. I found out you used signatures for marking accounts, but I don't understand the connection of this to the issue.
I can only tell about the accounts I have access to. Every transaction in Bank Hapoalim is returning the account balance after the transaction has been made, so it's quite easy to know the current balance of the account. The problem is that Hapoalim scraper doesn't support multiple accounts (is it planned?), but after it will implemented, it will be just there.
As for Leumi Card, this is actually problematic in some way. The website is providing the next month charge in the homepage, but it's the charge for all of your cards combined. In the transactions page, all of charges are detailed, but they are separated to ILS charges and foreign currency charges, and there are also postponed charges which makes the calculation more difficult. I chose to obtain the whole next month charge, and I couldn't find an easy way to get the month charges per account, but maybe this is the route we will choose to walk in and we should pursuit it, even if it somehow complicated.

**Comment by eshaham** [comment]
@chilikla sorry for not making it clearer.
what I mean by defining same signature/structure for all scrapers results, is thinking about the model we will be returning for the new account summary data. something along these lines:
```node
{
  success: boolean,
  accounts: [{
    accountNumber: string,
    summary: { // THIS IS THE NEW SUMMARY DATA
      balance: int,
      another_field: ...
    }
    txns: [...],  
  }],
  errorType: "invalidPassword"|"changePassword"|"timeout"|"generic", // only on success=false
  errorMessage: string, // only on success=false
}
```
Which other fields other than balance do you think provide value for account summary?

As for supporting multiple accounts in Bank Hapoalim, PR welcome 😄 
None of the bank scrapers support multiple accounts for now, but I believe it should be supported.

As far as I can recall, Leumi card shows a different table for each card, so we can probably take next month charge from there. I suspect we will be able to find this info for all credit cards scrapers.

Do you think the balance property should fit both bank current balance and credit cards next month's charge?

**Comment by chilikla** [comment]
Every financial account has some more information attached to the account itself, which can be useful, but maybe difficult to obtain.
Bank accounts, beside the balance property, may return the credit limit of the account, and maybe the interest rates for overdraft.
Credit cards, beside the next month charge (which we can call balance, I think), may return the credit limit, and how much it utilized until now.
Do you think we should gather more information?

As for your comment regarding Leumi Card and its tables, every account has 3 tables (in case of purchasing with foreign money) - table to ILS, table to foreign money which its charge in ILS will happened in the next month, and a table to to foreign money which its charge in ILS will happened in the months after this month. I guess we can skip this last table, and just should grab the summary of the 2 tables. It's feasible and I will work on it.
Obtaining credit limit in Leumi Card is also possible.

**Comment by eshaham** [comment]
@chilikla regarding account summary, sounds like a good start.
regarding Leumi Card, 👍 

I've created a new `account_summary` branch that we can use to iterate until all scrapers support account summary.
I think we should agree on a structure / signature for the account summary object before starting to push code into that branch.

**Comment by chilikla** [comment]
@eshaham, what you think of:
```
summary: {
  balance: int,
  credit_limit: int
}
```
for now?
The problem is with credit cards with "credit" accumulated through months. In this case, the credit use is not just the {credit_limit}-{balance}, and we can have {credit_use} by itself. Leumi Card provides this sum, but I don't know about other credit cards websites/apis. In bank accounts, anyway, the above equation is right (it's more {credit_limit}+{balance}, since the balance is negative), and there is no "credit" floating around messing with the final sum.
We can still add {credit_use} to bank accounts, even if the calculation is simple and obvious, just for the sake of consistency with credit cards, and then every account summary will be like this:
```
summary: {
  balance: int,
  credit_limit: int,
  credit_use: int
}
```

**Comment by eshaham** [comment]
@chilikla 
The direction sounds good.
I think all the numbers should be of type `double` and not `int`.
Also, to remain consistent, prop names should be lower camel case.

I believe that with credit cards, the balance should not only include the next upcoming payment, but any future payments. That reflects the true balance of the account, and will take care of the calculation issue you mentioned.

can we use the following?
```node
{
  ...
  summary: {
    balance: double,
    creditLimit: double,
    creditUsage: double,
  }
  ..
}
```

Allow me to iterate again and make sure I understand:

**banks**
balance: just the current account balance.
creditLimit: whatever it is for you account (מסגרת אשראי)
creditUsage: 0 if `balance` is positive, `creditLimit+balance` if `balance` is negative

**credit cards**
balance: amount due in any known future payments.
creditLimit: whatever it is for you account (מסגרת אשראי)
creditUsage: `creditLimit-balance`

**Comment by chilikla** [comment]
@eshaham 
The two suggestions are great, and I agreed this sums up the bank's `summary`.

As for `balance` for credit cards, your definition to it is indeed reflects the true balance of the account, but by that the balance property is not so much informative, since the `creditUsage` (I'm using the term `creditUtilization`, since I think is more accurate) reflecting the balance in a more accurate term. Besides that, I think there is an importance for providing the charge for the next month, and only the next month, as well.

I had the chance to work yesterday on the Leumi Card scraper, and I found some properties which relevant only to credit cards: `cardType` (Visa, Mastercard etc. as local companies branded credit cards as "Executive" in Leumi Card or so), and `chargeDate` for the date the credit card company withdraws your charges from your bank account. I'm not sure they supposed to be in summary, but I think they supposed to be somewhere.
Leumi Card has also rubric for next month charge in foreign currency, which for me is zero and I guess it's relevant for foreign currencies accounts. The charge is still in ILS, so I'm not sure what exactly this supposed to mean. Anyway, I used the properties names for ILS and FCY chrages in this structure: `cardIlsCharge`, `cardIlsDate`, `cardFcyCharge` and `cardFcyDate`.
Should we add `chargedCurrency` to all of these properties (I guess it will be same across all of the accounts summary data)?

So, for now I have this structure for Leumi Card account summary:
```
summary: {
  cardType: string,
  creditLimit: double,
  creditUtilization: double,
  cardIlsCharge: double,
  cardIlsDate: date, // in ISO format? they are just dates, without time, so maybe just strings?
  cardFcyCharge: double,
  cardFcyDate: date,
  chargedCurrency: string, // default is ILS
}
```

All of these properties are unique for credit cards, as far as I understand, and alongside the redundancy of the `balance` property, I'm not sure we should keep same signature across credit cards and bank accounts. Maybe you have another idea for the structure.

**Comment by chilikla** [comment]
Hi @eshaham,
I basically finished working on accounts summaries for the Leumi Card and Hapoalim scarpers, but I don't want to create a PR until, as you said, we will agree on the summary structure.
What do you think of the proposal I made in the last comment?

**Comment by eshaham** [comment]
@chilikla sorry, it has been a hectic week, and this week too (out on a business trip to the west coast).
`creditUtilization` indeed sounds much better than `creditUsage`, I now recollect using this name at a previous company.
Could you explain which value should `creditUtilization` field hold? Is it next month's charge? Or any pending unpaid balance?

I also agree about needing to create different summary structures for credit cards and banks. Which fields did you have in mind for bank accounts?

Regarding `cardType`, I'm not sure I'm not sure I'm following - I can think of 3 different things I would like to see in this field (these are just inaccurate examples):
1. `visa`, `mastercard`, `amex`, `diners`
2. `black`, `platinum`, `gold`, `international`
3. `executive`, `corporate`, `personal`
I'm not sure which one you meant, and field name should probably be more specific.

As for `chargedDate`, I think a better field would be `chargedDayOfMonth`, and it should hold an `int`.

I would change `cardIlsCharge` to `upcomingCardLocalCharge`, and `cardFcyCharge` to `upcomingCardForeginCharge`.

As for `chargedCurrency`, I don't think we need it for local transactions, but only for foreign ones. So I would change it to `upcomingCardForeignCurrency`.

final suggested structure for credit card summary:

```node
summary: {
  cardType: string, // need to see what to do with this field as mentioned above
  creditLimit: double,
  creditUtilization: double, 
  chargedDayOfMonth: int,
  upcomingCardLocalCharge: double,
  upcomingCardForeginCharge: double,
  upcomingCardForeginCurrency: string, // default is ILS
}
```

wdyt?

**Comment by chilikla** [comment]
@eshaham 
No need to apologize, I just thought you missed it. With most of your comments I agree, just few comments:

`creditUtilization`, as you guessed, indeed holds any pending unpaid balance. By including it, we will have information about the next payments as a whole, and not just the upcoming single month charge (which represented by `upcomingCardLocalCharge` and `upcomingCardForeginCharge`).

Regarding `cardType`, I referred to the definition the credit card company gave to the card, with its branding. Mine, for example, is 'לאומי קארד Multi מאסטרקארד'. I just thought to assign this string to the variable. It's possible to extract the credit card provider from this string, but as you can see, it's written in Hebrew, and I'm not sure it's important for something, but maybe it is. What do you think?

Regarding `upcomingCardForeginCurrency`, your comment is right, I just don't know how to handle it (that's why I commented about the default). In my Leumi Card's cards, even the `upcomingCardForeginCharge` charged in ILS, and I don't know if there are other options. I guess there are, I just don't have any idea how to assign to this property values other than ILS.

Regarding banks structure, your above mentioned suggestion is great:
```
{
  ...
  summary: {
    balance: double,
    creditLimit: double,
    creditUtilization: double,
  }
  ..
}
```
But I guess we should add the currency as `balanceCurrency` maybe?
Again I don't know how to handle foreign currencies there. Maybe we should have constant with common currency symbols as keys, and the currencies as values? What do you think?

Have a nice trip to the west coast :)

**Comment by eshaham** [comment]
> Regarding cardType, I referred to the definition the credit card company gave to the card, with its branding. Mine, for example, is 'לאומי קארד Multi מאסטרקארד'. I just thought to assign this string to the variable. It's possible to extract the credit card provider from this string, but as you can see, it's written in Hebrew, and I'm not sure it's important for something, but maybe it is. What do you think?

Ok, so this is different than what I had in mind. I don't find it as useful as the other options, but I definitely see value in this field. Perhaps we name it something like `cardName` or `cardTitle`?

> Regarding upcomingCardForeginCurrency, your comment is right, I just don't know how to handle it (that's why I commented about the default). In my Leumi Card's cards, even the upcomingCardForeginCharge charged in ILS, and I don't know if there are other options. I guess there are, I just don't have any idea how to assign to this property values other than ILS.

There are definitely other options - usually in credit cards issued by the bank, which allows you to setup a foreign checking account, and have all foreign transactions in that currency taken out of that account. I don't know what happens if you have transactions in other currencies, and when I think about it, not sure whether `upcomingCardForeginCurrency` can even be established. Problem is, `upcomingCardForeginCharge` is meaningless without it. So I'm kind of stuck here.

> But I guess we should add the currency as balanceCurrency maybe?
Again I don't know how to handle foreign currencies there. Maybe we should have constant with common currency symbols as keys, and the currencies as values? What do you think?

Not sure I understand what you mean here.

**Comment by chilikla** [comment]
@eshaham 

`cardName` sounds to me indeed accurate.

Regarding `upcomingCardForeginCharge` , in my Leumi Card balance page, the `upcomingCardForeginCharge` is empty, since all of my foreign currencies purchases are converted to ILS. In a case of existing `upcomingCardForeginCharge` and `upcomingCardForeginCurrency`, I can isolate both of them, just don't know what to do the unknown currency symbol.
I guess we can maintain a currencies object with currencies symbols and names, like you can find here:
https://gist.github.com/Fluidbyte/2973986.
What do you think?
This question is related to the third topic as well.

So I think maybe we can talk about this structure:
Banks:
```
{
  ...
  summary: {
    balance: double,
    creditLimit: double,
    creditUtilization: double,
    balanceCurrency: string,
  }
  ..
}
```
Credit cards:
```
{
  ...
  summary: {
    cardName: string,
    creditLimit: double,
    creditUtilization: double, 
    chargedDayOfMonth: int,
    upcomingCardLocalCharge: double,
    upcomingCardLocalCurrency: 'ILS',
    upcomingCardForeginCharge: double, // could be null
    upcomingCardForeginCurrency: string, // could be null
  }
  ...
}
```
Opinion?

**Comment by eshaham** [comment]
> In a case of existing upcomingCardForeginCharge and upcomingCardForeginCurrency, I can isolate both of them, just don't know what to do the unknown currency symbol.

@chilikla what I meant above is, unless the charge is in ILS, I don't think we can associate one currency for the entire charge, only for specific transactions. Because even if my card is setup to charge all USD transactions from my USD account, it could be that I'm also spending EUR, and these transactions will actually get converted to ILS.

What I suggest is, let's not tackle this at the moment - let's instead have the following:
`upcomingCardLocalCharge: double,`
~~`upcomingCardLocalCurrency: 'ILS',`~~ // we don't need it, as it is obvious
`upcomingCardForeginChargeInILS: double,` // either it is provided, or null

wdyt?

Regarding unknown currency symbol, did you run into anything so far? Let's leave it for the future, when someone runs into a real problem, and stick to major currencies for now.

**Comment by chilikla** [comment]
@eshaham I'm not exactly sure how it supposed to work, but in Leumi Card, for instance, there is separated entry for foreign currencies charge, like here:

![leumi card](https://user-images.githubusercontent.com/13035026/37622640-e1103de6-2bca-11e8-8dc3-cac1e4ebcfde.PNG)

I don't quite understand why my `upcomingCardForeignCharge` is also in ILS, but since all my abroad purchases are converted to ILS and listed under `upcomingCardLocalCharge` this doesn't bothers me, and I guess there is option for USD charges there, or EUR etc.
What I said is I can take the value from this field, and associate it with the foreign currency listed there.

> Regarding unknown currency symbol, did you run into anything so far? Let's leave it for the future, when someone runs into a real problem, and stick to major currencies for now.

I guess there are no unlimited number of currencies supported by the Israeli banks and credit cards, I just don't know the numerous currencies which indeed supported, that's why I suggested this.

Anyway, if you still think this is too complicated, I may agree with you and would prefer to move along and ditch the `upcomingCardForeignCharge` to `upcomingCardForeignChargeInILS`.

**Comment by eshaham** [comment]
@chilikla yes, let's try to simplify to these fields for now and see how it goes.
So, are we ready to get started? 😄

**Comment by chilikla** [comment]
Yeah, @eshaham 

So, this is the final structure:

Banks:
```
{
  ...
  summary: {
    balance: double,
    creditLimit: double,
    creditUtilization: double,
    balanceCurrency: string,
  }
  ..
}
```
Credit cards:
```
{
  ...
  summary: {
    cardName: string,
    creditLimit: double,
    creditUtilization: double, 
    chargedDayOfMonth: int,
    upcomingCardLocalCharge: double,
    upcomingCardForeignChargeInILS: double,
  }
  ...
}
```
Agreed?

**Comment by eshaham** [comment]
@chilikla yes, let's do this.
As I've written above, I've created a new `account_summary` branch that we can use to iterate until all scrapers support account summary.

**Comment by esakal** [comment]
@eshaham @chilikla I would like to add scraping of summary for bank Leumi. If I understand correctly I need to scrape the following fields:
```
summary: {
    balance: double,
    creditLimit: double,
    creditUtilization: double,
    balanceCurrency: 'ILS',
  }
```

## Balance
@eshaham It is easy to scrape the balance but I think It is better to do it after you will merge master into this branch (to extend the latest version of the Leumi scraper that includes the pending transactions support)

Also, since the balance value is different if you include/exclude pending transactions, I think we should consider adding a new option to the scrapers to allow the user to decide if to scrape pending transactions. eshaham/israeli-ynab-updater#27 is an example to such a use case, wdyt?

## creditUtilization and creditLimit
I looked in my account and since I don't have a credit defined I'm seeing the following content under the credit section.

![image](https://user-images.githubusercontent.com/4751797/38507401-14899bf8-3c25-11e8-9296-ea3407c6d9d6.png)

Do you have Leumi account? if so please send me the source for the following page (just fake the values):  `תפריט > עובר ושב >  פירוט מסגרות אשראי` and extend `מסגרת בעובר ושב`

In the meanwhile I will add scraping support for accounts that doesn't have credit defined.

**Comment by eshaham** [comment]
> @eshaham It is easy to scrape the balance but I think It is better to do it after you will merge master into this branch (to extend the latest version of the Leumi scraper that includes the pending transactions support)

Done.

> Also, since the balance value is different if you include/exclude pending transactions, I think we should consider adding a new option to the scrapers to allow the user to decide if to scrape pending transactions.

@esakal you're probably right, it is a good addition to the options, but I don't think pending transactions should affect this account summary and balance, so that should be irrelevant to this issue. If you agree, feel free to file an issue.

> I looked in my account and since I don't have a credit defined I'm seeing the following content under the credit section.

hmm, that's interesting. An account without a credit limit. @chilikla, I guess this should be mentioned in the README file. @esakal I don't have a Leumi bank account, but maybe @urigo or @ifeins could help.

**Comment by matanelgabsi** [comment]
Hi @chilikla @eshaham, is this still a work in progress? if no one is working on this, I would like to write the code to support bank hapoalim, leumi & otsar hachayal..

**Comment by esakal** [comment]
Hi @matanelgabsi this is issue is still in progress. @chilikla already created PRs for leumi card (#104) and bank Hapoalim (#103). They are both pending since April I will check them and try to push them forward. It would be great if you will support bank leumi and bank otsar hachayal.

**Comment by LiranBri** [comment]
Hi, is it possible now to get final balance for Habeleumi bank?
